syntax = "proto3";

package redapl;

import "timestamp.proto";
import "struct.proto";

// RawResourceV3 message
message RawResourceV3 {
    int64 org_id = 1;
    string type = 2;
    string name = 3;
    bytes fields_protobuf = 4;
    map<string, Value> fields_by_name = 5;
    Timestamp seen_at = 6;
    Timestamp expire_at = 7;
    InternalMetadata meta = 8;
    int64 version = 9;
    int64 tiebreaker = 10;
    string scope = 11;
}

// Metadata internal to the resource pipeline
message InternalMetadata {
	bool enable_debug_tracking = 1;
	ResourceStoreProxyMetadata resource_store_proxy = 2;
	string deduper_source = 3;
	Timestamp original_expire_at = 4 [deprecated=true];
	MigrationsMetadata migrations = 5;
	bool is_restreamed = 6;
	bool is_canary = 7;
}

message ResourceStoreProxyMetadata {
	string backend = 1;
	bool missing = 2;
	bool stale = 3;
	string err_message = 4;
	string provenance = 5;
	Timestamp written_at = 6;
}

// MigrationsMetadata is an internal type within the metadata to reflect a migrations applied to a resource
message MigrationsMetadata {
	repeated MigrationRecordMetadata records = 1;
	MigrationDerived derived = 2; // only exists on resources that were created from a derived transformation
	MigrationTestMetadata test_metadata = 3;
}

message MigrationTestMetadata {
	string original_resource_type = 1;
}

// MigrationRecordMetadata describes a migration applied to the current resource
message MigrationRecordMetadata {
	string revision_id = 1;
	Timestamp revision_created_at = 3;
	Timestamp timestamp = 4;
	MigrationBuildMetadata build_metadata = 5;
}

// MigrationDerived describes that the current resource was derived due to a migration from  a separate resource
message MigrationDerived {
	string from_resource_type = 1;
	string from_resource_name = 2;
	string from_revision_id = 3;
}

// MigrationBuildMetadata holds metadata for the migrator, when available, it provides additional information of the migration
// that was ran over the resource
message MigrationBuildMetadata {
	string build_commit = 1;
	Timestamp build_date = 2;
}